Users/jozessi/updateaml

dr-dolittle opened the pull request:
# Description

Update identity enablement from bool to actual object supporting user assigned managed identities

Add computes child resource

Update readmes

Add sections to bicep files

## Pipeline references

[![MachineLearningServices: Workspaces](https://github.com/quervernetzt/ResourceModules/actions/workflows/ms.machinelearningservices.workspaces.yml/badge.svg?branch=users%2Fjozessi%2Fupdateaml)](https://github.com/quervernetzt/ResourceModules/actions/workflows/ms.machinelearningservices.workspaces.yml)

# Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

MrMCake commented on arm/Microsoft.MachineLearningServices/workspaces/deploy.bicep:
```diff
@@ -31,8 +34,8 @@ param associatedContainerRegistryResourceId string = ''
-@description('Optional. Enables system assigned managed identity on the resource.')
-param systemAssignedIdentity bool = false
+@description('Optional. Identity for the resource.')
+param identity object = {}
```
Is there a reason why you changed this? We use the same implementation (almost) across the board and if the current one does not serve all use cases, we should update the other ones too. cc: @MariusStorhaug

dr-dolittle replied:
@MrMCake  @MariusStorhaug The way it was implemented it supported only system assigned managed identity. This way it is possible to provide user assigned managed identities as well. I propose to keep it this way (or similar) and if necessary also change it in the other modules. What do you think?

MariusStorhaug replied:
Could you check any other modules that supports user assigned identity, that would be where you see one way of supporting both. One thing we could work to improve, is the user assigned identity object that is created. Should really only be an array of strings.

dr-dolittle replied:
@MariusStorhaug I changed the implementation following your suggestion. Is it ok that way?

MariusStorhaug replied:
Ye, the implementation seems similar to the other ones now. I see in the diff that when we did change this last time, the resource type did not support `userAssignedIdentities`, now it seems it does. Thank you for adding it and help keeping the code consistent across modules!

MrMCake commented on arm/Microsoft.MachineLearningServices/workspaces/computes/deploy.bicep:
```diff
@@ -0,0 +1,118 @@
+param enableDefaultTelemetry bool = true
+
+// =========== //
+// Existing resources //
```
```suggestion
// Existing resource //
```
```suggestion
// Existing resources references // 
```

dr-dolittle replied:
Updated

MrMCake commented on arm/Microsoft.MachineLearningServices/workspaces/computes/deploy.bicep:
```diff
@@ -0,0 +1,118 @@
+    disableLocalAuth: computeDisableLocalAuth
+    computeType: computeType
+  }, (!empty(computeResourceId) ? {
+    resourceId: computeResourceId
```
If you'd agree we should try to stay aligned with the Azure resource reference. The rationale is, that it is (in theory) easier for consumers, as our readMe would be closer aligned to the offical Azure Resource Reference. If we diverge from it, there should be a good reason. 
> Note: This comment also relates to other properties like `computeDescription`

```suggestion
    resourceId: resourceId
```

dr-dolittle replied:
@MrMCake Updated. "description" is not possible -> "This expression is referencing its own declaration, which is not allowed.", therefore I use "computeDescription".

MrMCake replied:
@dr-dolittle actually it is. Use the annotation `sys.description()` instead. We have the same in case of e.g. the resources in the `Microsoft.Authorization` provider namespace

dr-dolittle replied:
@MrMCake Is updated.

MrMCake commented on arm/Microsoft.MachineLearningServices/workspaces/deploy.bicep:
```diff
@@ -153,6 +160,25 @@ resource workspace 'Microsoft.MachineLearningServices/workspaces@2021-04-01' = {
+    tags: tags
+    deployCompute: compute.deploy
+    computeLocation: compute.location
+    computeDescription: compute.description
```
If you specify it like this, `description` will become a required parameter even though the child module does not declare it as such.

The recommendation is (and the same accounts for all other properties): If a parameter is optional, use the syntax: `description: contains(compute, 'description') ? compute.description :  ''`
In words: If the compute object contains the property, use it, otherwise use the default specified in the child itself.

Unfortunately, there is no feature in bicep that it would automatically pick up the child default (like a special keyword)

dr-dolittle replied:
@MrMCake Updated.

MariusStorhaug commented:
Could you check any other modules that supports user assigned identity, that would be where you see one way of supporting both. One thing we could work to improve, is the user assigned identity object that is created. Should really only be an array of strings.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: dr-dolittle ***@***.***>
Sent: Thursday, April 14, 2022 11:18:35 PM
To: Azure/ResourceModules ***@***.***>
Cc: Marius Storhaug ***@***.***>; Mention ***@***.***>
Subject: Re: [Azure/ResourceModules] Users/jozessi/updateaml (PR #1298)


@dr-dolittle commented on this pull request.

________________________________

In arm/Microsoft.MachineLearningServices/workspaces/deploy.bicep<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2FResourceModules%2Fpull%2F1298%23discussion_r850811529&data=05%7C01%7Cmarius.storhaug%40microsoft.com%7C568da3f90039451a7daa08da1e5c5673%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637855679200440402%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=FXCZCoEv%2BdE%2BR5QTkZhmzTzDl6LxVd6tHEPMny0HL2A%3D&reserved=0>:

> @@ -31,8 +34,8 @@ param associatedContainerRegistryResourceId string = ''
 @description('Optional. Specify the type of lock.')
 param lock string = 'NotSpecified'

***@***.***('Optional. Enables system assigned managed identity on the resource.')
-param systemAssignedIdentity bool = false
***@***.***('Optional. Identity for the resource.')
+param identity object = {}


@MrMCake<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMrMCake&data=05%7C01%7Cmarius.storhaug%40microsoft.com%7C568da3f90039451a7daa08da1e5c5673%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637855679200440402%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=kgLXjxFAK28vY8UHBF79HDOquWuoezuByhtiB3mIUho%3D&reserved=0> @MariusStorhaug<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMariusStorhaug&data=05%7C01%7Cmarius.storhaug%40microsoft.com%7C568da3f90039451a7daa08da1e5c5673%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637855679200440402%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=I85i00RdMGhPCxJagp2wxwZaZd6nVSQHoktiblRfy9k%3D&reserved=0> The way it was implemented it supported only system assigned managed identity. This way it is possible to provide user assigned managed identities as well. I propose to keep it this way (or similar) and if necessary also change it in the other modules. What do you think?

—
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2FResourceModules%2Fpull%2F1298%23discussion_r850811529&data=05%7C01%7Cmarius.storhaug%40microsoft.com%7C568da3f90039451a7daa08da1e5c5673%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637855679200440402%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=FXCZCoEv%2BdE%2BR5QTkZhmzTzDl6LxVd6tHEPMny0HL2A%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAEHGXDLYZPOQC3GW7LH7R6TVFCDSXANCNFSM5TORCSVQ&data=05%7C01%7Cmarius.storhaug%40microsoft.com%7C568da3f90039451a7daa08da1e5c5673%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637855679200440402%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=vwYW%2FpwyXsqrq2JVBbUT94UnOcqM19AAXJxanq5iWb0%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

MariusStorhaug commented on arm/Microsoft.MachineLearningServices/workspaces/.parameters/parameters.json:
```diff
@@ -17,6 +17,34 @@
         },
+        "systemAssignedIdentity": {
+            "value": true
+        },
+        "computes": {
```
Lets add it to the deployment validation too:

```suggestion
        },
        "userAssignedIdentities": {
            "value": {
                "/subscriptions/<<subscriptionId>>/resourcegroups/validation-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/adp-<<namePrefix>>-az-msi-x-001": {}
            }
        },
        "computes": {
```

dr-dolittle replied:
@MariusStorhaug Is updated.

MariusStorhaug commented on arm/Microsoft.MachineLearningServices/workspaces/deploy.bicep:
```diff
@@ -153,6 +171,26 @@ resource workspace 'Microsoft.MachineLearningServices/workspaces@2021-04-01' = {
+    machineLearningWorkspaceName: workspace.name
+    name: compute.name
+    location: location
+    sku: sku
```
Do we want this to rely on the values from within the compute object that you iterate through or should it be inherited from the parent?

```suggestion
    sku: compute.sku
```

dr-dolittle replied:
@MariusStorhaug Updated.

MariusStorhaug commented on arm/Microsoft.MachineLearningServices/workspaces/deploy.bicep:
```diff
@@ -153,6 +171,26 @@ resource workspace 'Microsoft.MachineLearningServices/workspaces@2021-04-01' = {
+    name: compute.name
+    location: location
+    sku: sku
+    systemAssignedIdentity: systemAssignedIdentity
+    userAssignedIdentities: userAssignedIdentities
```
Do we want this to rely on the values from within the compute object that you iterate through or should it be inherited from the parent?

```suggestion
    systemAssignedIdentity: compute.systemAssignedIdentity
    userAssignedIdentities: compute.userAssignedIdentities
```

dr-dolittle replied:
@MariusStorhaug Updated.

MariusStorhaug commented on arm/Microsoft.MachineLearningServices/workspaces/.parameters/parameters.json:
```diff
@@ -17,6 +17,34 @@
+                    "description": "Default CPU Cluster",
+                    "disableLocalAuth": false,
+                    "deploy": true,
+                    "type": "AmlCompute",
```
Lets discuss this point under the suggestion on the deploy.bicep file for the workspaces module where you call the compute module. Come back here if you want to treat these independent from the parent settings.

```suggestion
                    "type": "AmlCompute",
                    "systemAssignedIdentity": true,
                    "userAssignedIdentities": {
                        "/subscriptions/<<subscriptionId>>/resourcegroups/validation-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/adp-<<namePrefix>>-az-msi-x-001": {}
                    },
```

dr-dolittle replied:
@MariusStorhaug Updated.

eriqua commented on arm/Microsoft.MachineLearningServices/workspaces/computes/deploy.bicep:
```diff
@@ -0,0 +1,132 @@
+output resourceId string = machineLearningWorkspaceCompute.id
+
+@description('The resource group the compute was deployed into.')
+output resourceGroupName string = resourceGroup().name
```
I'd suggest to add this output and regenerate the readme file
```suggestion
output resourceGroupName string = resourceGroup().name

@description('The principal ID of the system assigned identity.')
output systemAssignedPrincipalId string = systemAssignedIdentity && contains(machineLearningWorkspaceCompute.identity, 'principalId') ? machineLearningWorkspaceCompute.identity.principalId : ''
```

dr-dolittle replied:
@eriqua Is updated.

dr-dolittle commented:
@MrMCake @MariusStorhaug @eriqua Thanks for your feedback. I implemented your changes as suggested. There are two more open topics: in case of "create" it is actually possible to deploy multiple times without error, in case of attach it fails. Also, I want to clarify which properties are actually required for attach versus create. I will update it as soon as I know more.

dr-dolittle commented:
> @MrMCake @MariusStorhaug @eriqua Thanks for your feedback. I implemented your changes as suggested. There are two more open topics: in case of "create" it is actually possible to deploy multiple times without error, in case of "attach" it fails. Also, I want to clarify which properties are actually required for "attach" versus "create". I will update it as soon as I know more.

@MrMCake @MariusStorhaug @eriqua Is now adjusted and fine from my perspective. Please review.

But the pipeline fails dues to failing static validation -> https://github.com/quervernetzt/ResourceModules/actions/runs/2179547321:

<img width="911" alt="image" src="https://user-images.githubusercontent.com/26380370/163709088-19637bcf-a0df-43b0-8c8d-758167859119.png">
<img width="976" alt="image" src="https://user-images.githubusercontent.com/26380370/163709096-174e2eb1-93b6-48a5-9bed-033125745f2a.png">

dr-dolittle commented:
> > @MrMCake @MariusStorhaug @eriqua Thanks for your feedback. I implemented your changes as suggested. There are two more open topics: in case of "create" it is actually possible to deploy multiple times without error, in case of "attach" it fails. Also, I want to clarify which properties are actually required for "attach" versus "create". I will update it as soon as I know more.
> 
> @MrMCake @MariusStorhaug @eriqua Is now adjusted and fine from my perspective. Please review.
> 
> But the pipeline fails dues to failing static validation -> https://github.com/quervernetzt/ResourceModules/actions/runs/2179547321:
> 
> <img alt="image" width="911" src="https://user-images.githubusercontent.com/26380370/163709088-19637bcf-a0df-43b0-8c8d-758167859119.png">
> 
> <img alt="image" width="976" src="https://user-images.githubusercontent.com/26380370/163709096-174e2eb1-93b6-48a5-9bed-033125745f2a.png">
> 
> Can you please check?
> 
> Thanks

@MrMCake This one also fails -> https://github.com/quervernetzt/ResourceModules/actions/runs/2189802004

![image](https://user-images.githubusercontent.com/26380370/164012487-6526b259-d9a4-4ddb-a0c9-d6cd8d87c5cf.png)

dr-dolittle commented:
@MrMCake Pipeline run was successful. Can we merge now? Thanks

MrMCake commented on arm/Microsoft.MachineLearningServices/workspaces/.parameters/parameters.json:
```diff
@@ -17,6 +17,53 @@
+                    }
+                }
+                // {
+                //     "name": "DefaultAKS",
```
I guess this is commented because it cannot be deployed in parallel?

dr-dolittle replied:
@MrMCake The reason why it is commented is to show how it would work but it is not enabled as it would require an AKS cluster. Do we have an AKS as dependency available?

It is possible to deploy multiple compute resources in parallel.

MrMCake replied:
Not yet, no. But we can do this in the future with a new dependency approach. I'm usually not a big fan of these kind of comments - so could you please add a smal disclaimer on top if it saying : `This is an example of a link to an existing AKS resource`

dr-dolittle replied:
@MrMCake I removed the commented part.

MrMCake commented on arm/Microsoft.MachineLearningServices/workspaces/computes/deploy.bicep:
```diff
@@ -0,0 +1,136 @@
+  }
+}
+
+resource machineLearningWorkspaceCompute 'Microsoft.MachineLearningServices/workspaces/computes@2022-01-01-preview' = if (deployCompute == true) {
```
Why the condition?

dr-dolittle replied:
@MrMCake As written in the README doing a deployment for an attached compute (e.g. AKS) multiple times will fail, i.e. the first one will succeed and attach the resource, but as the operation is not idempotent subsequent resources will fail. So, when you attach a compute you need to set the flag to false after the first deployment. For created resources via AML this flag is not relevant as the operation is idempotent. The PG is aware of the problem but so far has not resolved it.

MrMCake replied:
ah right, my bad
